h4fromh5.1
.\" Copyright (c) 2005 Massachusetts Institute of Technology
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be
.\" included in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\" EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
.\" CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
.\" TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
.\" SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.TH H4FROMH5 1 "September 2, 2005" "h5utils" "h5utils"
.SH NAME
h4fromh5 \- convert HDF5 scientific dataset to an HDF4 file
.SH SYNOPSIS
.B h4fromh5
[\fIOPTION\fR]... [\fIHDF4FILE\fR]...
.SH DESCRIPTION
.PP
." Add any additional description here
h4fromh5 takes one or more files in HDF5 format and outputs files in
HDF4 format containing the datasets from the HDF5 files.  (Currently,
only a single dataset per HDF5 file is converted.)

HDF4 and HDF5 are free, portable binary formats and supporting libraries
developed by the National Center for Supercomputing Applications at
the University of Illinois in Urbana-Champaign.

 A single
.I h5
file can contain multiple data sets; by default,
.I h4fromh5
converts the first dataset, but this can be changed via the
.B -d
option, or by using the syntax \fIHDF5FILE:DATASET\fR.

The most basic usage is something like \'h4fromh5 foo.h5\', which
will output a file foo.hdf containing the scientific dataset from
foo.h5.
.SH OPTIONS
.TP
.B -h
Display help on the command-line options and usage.
.TP
.B -V
Print the version number and copyright info for h4fromh5.
.TP
.B -v
Verbose output.
.TP
.B -T
Transpose the output dataset (e.g. LxMxN becomes NxMxL).  This is often
useful because HDF5 programs typically follow C (row-major) conventions
while HDF4 programs often follow Fortran (column-major, transposed)
conventions for array ordering.
.TP
\fB\-o\fR \fIfile\fR
Send HDF output to
.I file
rather than to the input filename with .h5 replaced with .hdf (the
default).
.TP
\fB\-d\fR \fIname\fR
Read from dataset
.I name
in the input; otherwise, the first dataset in the input file is used.
Alternatively, use the syntax \fIHDF5FILE:DATASET\fR when the
input file names are specified.
.SH BUGS
Send bug reports to S. G. Johnson, stevenj@alum.mit.edu.
.SH AUTHORS
Written by Steven G. Johnson.  Copyright (c) 2005 by the Massachusetts
Institute of Technology.


NEWS
h5utils 1.10 (9/2/05)

  * Added h4fromh5 utility.  (NCSA seems to be no longer shipping
    the HDF5/HDF4 conversion tools with the latest HDF5 release.)

  * Added dkbluered color map, which is similar to bluered but
    uses a somewhat wider range of colors.

h5utils 1.9.1 (8/5/04)

  * Fix h5topng compilation failure with some non-C99 compilers;
    thanks to Maarten van Reeuwijk for the bug report.

h5utils 1.9 (7/12/04)

  * Added new h5math utility, which creates and combines HDF5 datasets
    using a user-specified mathematical expression.  (Requires GNU
    libmatheval to be installed.)

  * h5topng: Matlab-like start:end or start:step:end notation for
    slice indices, to allow a whole sequence of slices to be output as
    a sequence of PNG images.

  * h5topng: if contour/overlay dataset does not have same dimensions
    as output data, it is periodically "tiled" over the output.

h5utils 1.8 (5/22/04)

  * New -A and -a options for h5topng to allow translucent overlays
    from one file onto another (an alternative or complement to the
    -C contour-overlay option).

  * h5topng uses 24-bit direct color by default (use -8 option for
    old 8-bit behavior).

  * h5topng uses columns/rows for x/y by default (use -T to swap),
    the opposite of the old behavior.

  * There is no default -z 0 slice dimension in h5topng/h5totxt any more.
    You must specify a slice for 3+ dimensional data in h5topng.
    h5totxt dumps the whole data file by default unless one or more
    slices are specified.

  * Support specifying multiple slice dimensions for 4+ dimensional
    datasets, with new "-t" option to indicate final dimension.

  * Slices are also supported now in h5tovtk and h5tov5d.

  * New -. option in h5totxt to specify number of significant digits; 
    output 16 digits by default instead of 6, previously.

h5utils 1.7.1 (3/16/02)

  * Fixed array overrun in h5topng that caused a floating-point exception
    on Alphas; thanks to Marin Soljacic for the bug report.

h5utils 1.7 (3/9/02)

  * h5topng now supports multiple, user-definable color tables, a number
    of which are provided.  INCOMPATIBLE CHANGE: the -c option now has
    the syntax: -c <colortable>.  The old behavior corresponds to the
    included "bluered" colortable, invoked via: -c bluered.

  * New -R option for h5topng to use a consistent color scale for
    all specified files.

  * New -0 option for h5topng and h5totxt that shifts the origin of the
    slice coordinates to the dataset center.

  * Added h5tovtk program to output VTK (Visualization ToolKit) data files.

  * Support -T (transpose dimensions) option in h5tov5d.

  * Fixed bug in h5topng that caused extra rows/columns of garbage pixels
    to be written at the edges of images when scaling was used.

  * When compiling the h5read Octave plugin, respect the CPPFLAGS and LDFLAGS 
    environment variables.  Thanks to Max Colice for the bug report.

  * Fixed problem when --without-h5tov5d and --without-h5fromh4 are used. 
    Thanks to Nikola Ivanov Nikolov for the fix.

h5utils 1.6 (1/17/01)

  * Don't build h5fromh4 if the superior h4toh5 tool (from HDF5 1.4)
    is present.  Also added --{with,without}-h5fromh4 option to
    configure to force whether h5fromh4 is built.

h5utils 1.5.1 (12/9/00)

  * Support manually disabling Octave plugin support
    (configure --without-octave) in case of C++ problems.

  * Support Vis5d+ and Debian HDF header file locations.

h5utils 1.5 (7/9/00)

  * Added h5fromh4 program to convert HDF4 datasets to HDF5 format.

  * Added -S <s> option to h5topng as a shortcut for -X <s> -Y <s>.

h5utils 1.4 (5/28/00)

  * Added h5fromtxt program to convert text input to an HDF5 dataset.

h5utils 1.3.4 (1/31/00)

  * Improved -C contour plotting in h5topng.

  * Fix in h5topng man page (thanks to Christoph Becher).

h5utils 1.3.3 (1/30/00)

  * Bug fix in h5topng (would sometimes output solid black images).
    Thanks to Karl Koch for the bug report.

h5utils 1.3.2 (1/28/00)

  * Added h5topng -Z option to center color scale on zero.

  * Now support h5topng -C filename:dataset.

h5utils 1.3.1 (1/27/00)

  * Bug fixes in dataset name-handling, especially when using h5tov5d to
    join multiple datasets into one output file.

h5utils 1.3 (1/21/00)

  * You can now specify individual datasets within a file by using
    <filename>:<dataset> instead of just <filename> with h5topng,
    h5totxt, and h5tov5d.

h5utils 1.2.3: (1/20/00)

  * Fixed minor bug in 'h5totxt -h'.

h5utils 1.2.2: (1/12/00)

  * Makefile now includes CPPFLAGS in the compiler flags, making it
    easier to use header files in non-standard locations.  (CPPFLAGS
    is the proper place to put -I flags for the configure script.)

h5utils 1.2.1:  (1/1/00)

  * Modified -o option of h5tov5d to join datasets into a single Vis5d file.

h5utils 1.2:  (12/31/99)

  * Added h5tov5d program for converting to Vis5d format.

  * Improved installation; h5read.oct now goes into the sitewide Octave
    plugins directory, and things work correctly when only some of the
    utilities are compiled.

h5utils 1.1:  (12/6/99)

  * Added h5totxt program for exporting 2d slices of HDF5 files to text
    suitable for importing into a spreadsheet.

h5utils 1.0:  (11/22/99)

  * Initial release.


h4fromh5.c
/* Copyright (c) 2005 Massachusetts Institute of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <unistd.h>

#include "config.h"
#include "arrayh5.h"
#include "arrayh4.h"
#include "copyright.h"
#include "h5utils.h"

#define CHECK(cond, msg) { if (!(cond)) { fprintf(stderr, "h4fromh5 error: %s\n", msg); exit(EXIT_FAILURE); } }

void usage(FILE *f)
{
     fprintf(f, "Usage: h4fromh5 [options] [<hdf4-files>]\n"
	     "Options:\n"
	     "         -h : this help message\n"
             "         -V : print version number and copyright\n"
	     "         -v : verbose output\n"
	     "         -T : transposed output\n"
	     "  -o <file> : output to HDF4 file <file>\n"
	     "  -d <name> : use dataset <name> in the input file\n"
	     "              -- you can also specify a dataset via <file>:<name>\n"
	  );
}

int main(int argc, char **argv)
{
     char *h4_fname = NULL;
     char *data_name = NULL;
     extern char *optarg;
     extern int optind;
     int c;
     int ifile;
     int verbose = 0, transpose = 0;

     while ((c = getopt(argc, argv, "hd:vTo:V")) != -1)
	  switch (c) {
	      case 'h':
		   usage(stdout);
		   return EXIT_SUCCESS;
	      case 'V':
		   printf("h4fromh5 " PACKAGE_VERSION " by Steven G. Johnson\n" 
			  COPYRIGHT);
		   return EXIT_SUCCESS;
	      case 'v':
		   verbose = 1;
		   break;
	      case 'T':
		   transpose = 1;
		   break;
	      case 'd':
		   free(data_name);
		   data_name = my_strdup(optarg);
		   break;		   
	      case 'o':
		   free(h4_fname);
		   h4_fname = my_strdup(optarg);
		   break;
	      default:
		   fprintf(stderr, "Invalid argument -%c\n", c);
		   usage(stderr);
		   return EXIT_FAILURE;
	  }
     if (optind == argc) {  /* no parameters left */
	  usage(stderr);
	  return EXIT_FAILURE;
     }

     if (h4_fname && optind + 1 < argc) {
	  fprintf(stderr, "h4fromh5: only one .h5 file can be used with -o\n");
	  return EXIT_FAILURE;
     }

     for (ifile = optind; ifile < argc; ++ifile) {
	  char *h5_fname, *dname;
	  arrayh4 a4;
	  int i, err;
	  int32 dims_copy[ARRAYH4_MAX_RANK];
	  char *cur_h4_fname = h4_fname;
	  arrayh5 a;

	  h5_fname = split_fname(argv[ifile], &dname);
	  if (!dname[0])
               dname = data_name;

	  if (!cur_h4_fname)
	       cur_h4_fname = replace_suffix(h5_fname, ".h5", ".hdf");

	  if (verbose)
	       printf("Reading HDF5 input file \"%s\"...\n", h5_fname);
	  err = arrayh5_read(&a, h5_fname, dname, NULL, 0, 0, 0, 0);
	  CHECK(!err, arrayh5_read_strerror[err]);

	  if (transpose)
	       arrayh5_transpose(&a);

	  CHECK(a.rank <= ARRAYH4_MAX_RANK, "HDF5 rank is too big");
	  for (i = 0; i < a.rank; ++i)
	       dims_copy[i] = a.dims[i];

	  CHECK(arrayh4_create(&a4, DFNT_FLOAT64, a.rank, dims_copy),
		"error allocating HDF4 data");
	  
	  for (i = 0; i < a.N; ++i)
	       a4.p.d[i] = a.data[i];

	  if (verbose) {
	       double a_min, a_max;
	       arrayh5_getrange(a, &a_min, &a_max);
	       printf("data ranges from %g to %g.\n", a_min, a_max);
	  }
	  
	  if (verbose) {
	       int i;
	       printf("Writing size %d", a.dims[0]);
	       for (i = 1; i < a.rank; ++i)
		    printf("x%d", a.dims[i]);
	       printf(" data to %s\n", cur_h4_fname);
	  }

	  arrayh5_destroy(a);

	  arrayh4_write(cur_h4_fname, a4);
	  arrayh4_destroy(a4);

	  if (h4_fname != cur_h4_fname)
	       free(cur_h4_fname);
	  free(h5_fname);
     }

     return EXIT_SUCCESS;
}
